File: apps/relayd/internal/relay/fabric_envelope.go
package relay

import (
	"encoding/binary"
	"fmt"
)

const (
	FabricProtocolVersion = 3
	fabricEnvelopeHeader  = 24
)

const (
	FabricKindLinkOpen   = "link_open"
	FabricKindLinkReady  = "link_ready"
	FabricKindData       = "relay_data_frame"
	FabricKindPeerClosed = "relay_peer_closed"
	FabricKindError      = "relay_error"
)

// FabricEnvelope is the v3 outer frame. The relay learns only Fabric, Node,
// Link and stream routing fields; payload stays end-to-end encrypted.
type FabricEnvelope struct {
	Version  int
	FabricID string
	NodeID   string
	LinkID   string
	StreamID string
	Seq      uint32
	Ack      uint32
	Kind     string
	Priority string
	Payload  []byte
}

func ParseFabricEnvelopeView(data []byte, maxBytes int64) (FabricEnvelope, error) {
	if maxBytes > 0 && int64(len(data)) > maxBytes {
		return FabricEnvelope{}, ErrFrameTooLarge
	}
	if len(data) < fabricEnvelopeHeader || int(data[0]) != FabricProtocolVersion {
		return FabricEnvelope{}, ErrInvalidEnvelope
	}
	kind, ok := fabricKindFromCode(data[1])
	if !ok {
		return FabricEnvelope{}, ErrInvalidEnvelope
	}
	priority, ok := priorityFromCode(data[2])
	if !ok {
		return FabricEnvelope{}, ErrInvalidEnvelope
	}
	fabricLen := int(binary.BigEndian.Uint16(data[4:6]))
	nodeLen := int(binary.BigEndian.Uint16(data[6:8]))
	linkLen := int(binary.BigEndian.Uint16(data[8:10]))
	streamLen := int(binary.BigEndian.Uint16(data[10:12]))
	seq := binary.BigEndian.Uint32(data[12:16])
	ack := binary.BigEndian.Uint32(data[16:20])
	payloadLen := int(binary.BigEndian.Uint32(data[20:24]))
	expected := fabricEnvelopeHeader + fabricLen + nodeLen + linkLen + streamLen + payloadLen
	if fabricLen == 0 || nodeLen == 0 || linkLen == 0 || payloadLen == 0 || expected != len(data) {
		return FabricEnvelope{}, ErrInvalidEnvelope
	}
	offset := fabricEnvelopeHeader
	env := FabricEnvelope{
		Version: FabricProtocolVersion, FabricID: string(data[offset : offset+fabricLen]),
		Seq: seq, Ack: ack, Kind: kind, Priority: priority,
	}
	offset += fabricLen
	env.NodeID = string(data[offset : offset+nodeLen])
	offset += nodeLen
	env.LinkID = string(data[offset : offset+linkLen])
	offset += linkLen
	env.StreamID = string(data[offset : offset+streamLen])
	offset += streamLen
	env.Payload = data[offset:]
	if err := env.Validate(maxBytes); err != nil {
		return FabricEnvelope{}, err
	}
	return env, nil
}

func EncodeFabricEnvelope(env FabricEnvelope, maxBytes int64) ([]byte, error) {
	if env.Priority == "" {
		env.Priority = PriorityControl
	}
	if err := env.Validate(maxBytes); err != nil {
		return nil, err
	}
	kind, _ := fabricKindCode(env.Kind)
	priority, _ := priorityCode(env.Priority)
	out := make([]byte, fabricEnvelopeHeader+len(env.FabricID)+len(env.NodeID)+len(env.LinkID)+len(env.StreamID)+len(env.Payload))
	out[0], out[1], out[2] = FabricProtocolVersion, kind, priority
	binary.BigEndian.PutUint16(out[4:6], uint16(len(env.FabricID)))
	binary.BigEndian.PutUint16(out[6:8], uint16(len(env.NodeID)))
	binary.BigEndian.PutUint16(out[8:10], uint16(len(env.LinkID)))
	binary.BigEndian.PutUint16(out[10:12], uint16(len(env.StreamID)))
	binary.BigEndian.PutUint32(out[12:16], env.Seq)
	binary.BigEndian.PutUint32(out[16:20], env.Ack)
	binary.BigEndian.PutUint32(out[20:24], uint32(len(env.Payload)))
	offset := fabricEnvelopeHeader
	for _, field := range []string{env.FabricID, env.NodeID, env.LinkID, env.StreamID} {
		copy(out[offset:], field)
		offset += len(field)
	}
	copy(out[offset:], env.Payload)
	return out, nil
}

func (env FabricEnvelope) Validate(maxBytes int64) error {
	if env.Version != FabricProtocolVersion || env.FabricID == "" || env.NodeID == "" || env.LinkID == "" || env.Kind == "" || env.Priority == "" || len(env.Payload) == 0 {
		return ErrInvalidEnvelope
	}
	if _, ok := fabricKindCode(env.Kind); !ok {
		return ErrInvalidEnvelope
	}
	if _, ok := priorityCode(env.Priority); !ok {
		return ErrInvalidEnvelope
	}
	for _, value := range []string{env.FabricID, env.NodeID, env.LinkID, env.StreamID} {
		if len(value) > 0xffff {
			return fmt.Errorf("%w: identifier too long", ErrInvalidEnvelope)
		}
	}
	if maxBytes > 0 && int64(fabricEnvelopeHeader+len(env.FabricID)+len(env.NodeID)+len(env.LinkID)+len(env.StreamID)+len(env.Payload)) > maxBytes {
		return ErrFrameTooLarge
	}
	return nil
}

func fabricKindCode(kind string) (byte, bool) {
	switch kind {
	case FabricKindLinkOpen:
		return 1, true
	case FabricKindLinkReady:
		return 2, true
	case FabricKindData:
		return 3, true
	case FabricKindPeerClosed:
		return 4, true
	case FabricKindError:
		return 5, true
	default:
		return 0, false
	}
}

func fabricKindFromCode(code byte) (string, bool) {
	for _, kind := range []string{FabricKindLinkOpen, FabricKindLinkReady, FabricKindData, FabricKindPeerClosed, FabricKindError} {
		if candidate, ok := fabricKindCode(kind); ok && candidate == code {
			return kind, true
		}
	}
	return "", false
}

File: apps/server/src/modules/relay-transport/fabric-envelope.ts
import { AppError } from '../../errors/app-error'
import type { RelayEnvelope, RelayEnvelopeKind, RelayPriority } from './protocol'
import { RELAY_PROTOCOL_VERSION } from './protocol'
import type { RelayOutboundEnvelope } from './session'

export const FABRIC_ENVELOPE_VERSION = 3
const HEADER_BYTES = 24

export interface FabricEnvelopeRoute { fabricId: string, nodeId: string, linkId: string }
export interface DecodedFabricEnvelope extends FabricEnvelopeRoute { streamId?: string, seq: number, ack: number, kind: 'link_open' | 'link_ready' | RelayEnvelopeKind, priority: RelayPriority, payload: Uint8Array }

const kindToCode = { link_open: 1, link_ready: 2, relay_data_frame: 3, relay_peer_closed: 4, relay_error: 5 } as const
const codeToKind = new Map<number, DecodedFabricEnvelope['kind']>(Object.entries(kindToCode).map(([kind, code]) => [code, kind as DecodedFabricEnvelope['kind']]))

export function encodeFabricEnvelope(route: FabricEnvelopeRoute, frame: RelayOutboundEnvelope): Uint8Array {
  const fields = [utf8(route.fabricId, 'Fabric id'), utf8(route.nodeId, 'Node id'), utf8(route.linkId, 'Link id'), frame.streamId ? utf8(frame.streamId, 'Stream id') : new Uint8Array()]
  const kind = kindToCode[frame.kind]
  if (!kind || frame.payload.byteLength === 0) { throw protocolError('Invalid Fabric relay envelope.') }
  const out = new Uint8Array(HEADER_BYTES + fields.reduce((total, field) => total + field.byteLength, 0) + frame.payload.byteLength)
  const view = new DataView(out.buffer, out.byteOffset, out.byteLength)
  out[0] = FABRIC_ENVELOPE_VERSION; out[1] = kind; out[2] = frame.priority === 'control' ? 1 : 2
  view.setUint16(4, fields[0].byteLength); view.setUint16(6, fields[1].byteLength); view.setUint16(8, fields[2].byteLength); view.setUint16(10, fields[3].byteLength)
  view.setUint32(12, frame.seq); view.setUint32(16, 0); view.setUint32(20, frame.payload.byteLength)
  let offset = HEADER_BYTES
  for (const field of fields) { out.set(field, offset); offset += field.byteLength }
  out.set(frame.payload, offset)
  return out
}

export function decodeFabricEnvelope(bytes: Uint8Array): DecodedFabricEnvelope {
  if (bytes.byteLength < HEADER_BYTES || bytes[0] !== FABRIC_ENVELOPE_VERSION) { throw protocolError('Unsupported Fabric relay envelope.') }
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength)
  const kind = codeToKind.get(bytes[1]); const priority = bytes[2] === 1 ? 'control' : bytes[2] === 2 ? 'data' : null
  const lengths = [view.getUint16(4), view.getUint16(6), view.getUint16(8), view.getUint16(10)]
  const payloadLength = view.getUint32(20)
  const expected = HEADER_BYTES + lengths.reduce((total, length) => total + length, 0) + payloadLength
  if (!kind || !priority || !lengths[0] || !lengths[1] || !lengths[2] || !payloadLength || expected !== bytes.byteLength) { throw protocolError('Invalid Fabric relay envelope.') }
  let offset = HEADER_BYTES
  const [fabricId, nodeId, linkId, streamId] = lengths.map((length, index) => { const result = length ? decode(bytes.subarray(offset, offset + length), ['fabric id', 'node id', 'link id', 'stream id'][index]!) : undefined; offset += length; return result })
  return { fabricId: fabricId!, nodeId: nodeId!, linkId: linkId!, ...(streamId ? { streamId } : {}), seq: view.getUint32(12), ack: view.getUint32(16), kind, priority, payload: bytes.subarray(offset) }
}

/**
 * Converts a real v3 envelope into the logical envelope consumed by the
 * existing encrypted session state machine. Routing has already been checked
 * by FabricNodeLinkManager, so it is deliberately not retained here.
 */
export function toRelaySessionEnvelope(envelope: DecodedFabricEnvelope): RelayEnvelope {
  if (envelope.kind === 'link_open' || envelope.kind === 'link_ready') { throw protocolError('Fabric link control is not an encrypted session frame.') }
  return { version: RELAY_PROTOCOL_VERSION, roomId: envelope.linkId, seq: envelope.seq, kind: envelope.kind, priority: envelope.priority, ...(envelope.streamId ? { streamId: envelope.streamId } : {}), payload: envelope.payload }
}

function utf8(value: string, label: string): Uint8Array { const bytes = new TextEncoder().encode(value); if (!bytes.byteLength || bytes.byteLength > 0xFFFF) { throw protocolError(`${label} is invalid.`) } return bytes }
function decode(value: Uint8Array, label: string): string {
 try { return new TextDecoder('utf-8', { fatal: true }).decode(value) }
 catch { throw protocolError(`${label} is invalid.`) }
}
function protocolError(message: string): AppError { return new AppError({ code: 'fabric_protocol_invalid_frame', status: 400, message }) }
